testing-common/src/main/java/org/hypertrace/agent/testing/AbstractHttpClientTest.java

```diff
@@ -0,0 +1,264 @@
+    } else {
+      Assertions.assertEquals(1, traces.get(0).size());
+      SpanData clientSpan = traces.get(0).get(0);
+      assertEchoUrlEncoded(clientSpan, body);
+    }
+  }
+
+  @Test
+  public void echoPlainText()
+      throws TimeoutException, InterruptedException, IOException, ExecutionException {
+    String body = "foobar";
+    String uri = String.format(ECHO_PATH_FORMAT, testHttpServer.port());
+
+    Pair<Integer, String> statusBodyPair = doPostRequest(uri, headers, body, "text/plain");
+
+    Assertions.assertEquals(200, statusBodyPair.getLeft());
+    Assertions.assertEquals(body, statusBodyPair.getRight());
+
+    TEST_WRITER.waitForTraces(1);
+    List<List<SpanData>> traces = TEST_WRITER.getTraces();
+    Assertions.assertEquals(1, traces.size());
+    Assertions.assertEquals(1, traces.get(0).size());
+    SpanData clientSpan = traces.get(0).get(0);
+
+    assertNullBodies(clientSpan);
+  }
+
+  @Test
+  public void getNoContent()
+      throws IOException, TimeoutException, InterruptedException, ExecutionException {
+    String uri = String.format(GET_NO_CONTENT_PATH_FORMAT, testHttpServer.port());
+
+    Pair<Integer, String> statusBodyPair = doGetRequest(uri, headers);
+
+    Assertions.assertEquals(204, statusBodyPair.getLeft());
+    Assertions.assertNull(statusBodyPair.getRight());
+
+    TEST_WRITER.waitForTraces(1);
+    List<List<SpanData>> traces = TEST_WRITER.getTraces();
+    Assertions.assertEquals(1, traces.size());
+    Assertions.assertEquals(1, traces.get(0).size());
+    SpanData clientSpan = traces.get(0).get(0);
+
+    assertNullBodies(clientSpan);
+  }
+
+  @Test
+  public void getJson()
+      throws IOException, TimeoutException, InterruptedException, ExecutionException {
+    String uri = String.format(GET_JSON_PATH_FORMAT, testHttpServer.port());
+
+    Pair<Integer, String> statusBodyPair = doGetRequest(uri, headers);
+
+    Assertions.assertEquals(200, statusBodyPair.getLeft());
+    Assertions.assertEquals(TestHttpServer.GetJsonHandler.RESPONSE_BODY, statusBodyPair.getRight());
+
+    TEST_WRITER.waitForTraces(1);
+    List<List<SpanData>> traces = TEST_WRITER.getTraces();
+    Assertions.assertEquals(1, traces.size());
+
+    if (hasResponseBodySpan) {
+      Assertions.assertEquals(2, traces.get(0).size());
+      SpanData clientSpan = traces.get(0).get(0);
+      SpanData responseBodySpan = traces.get(0).get(1);
+      Assertions.assertNull(
+          clientSpan.getAttributes().get(HypertraceSemanticAttributes.HTTP_REQUEST_BODY));
+      Assertions.assertEquals(
+          TestHttpServer.GetJsonHandler.RESPONSE_BODY,
+          responseBodySpan.getAttributes().get(HypertraceSemanticAttributes.HTTP_RESPONSE_BODY));
+    } else {
+      Assertions.assertEquals(1, traces.get(0).size());
+      SpanData clientSpan = traces.get(0).get(0);
+      Assertions.assertNull(
+          clientSpan.getAttributes().get(HypertraceSemanticAttributes.HTTP_REQUEST_BODY));
```

Comment:
dry the branches, e.g. this is is asserted for both branches. Each branch should have only the code that is unique for a branch.